hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java

```diff
@@ -522,7 +489,75 @@ private void handleRetry(Object cause) {
         }
     }
 
-    boolean checkBackupTimeout(long timeoutMillis) {
+    /**
+     * Checks if this Invocation has received a heart-beat in time.
+     *
+     * If the response is already set, or if a heart-beat has been received in time, then the false is returned.
+     *
+     * If no heart-beat has been received, then the future.set is called with NO_HEARTBEAT
+     * and true is returned.
+     *
+     * gets called from the monitor-thread
+     *
+     * @return true if there is a timeout detected, false otherwise.
+     */
+    boolean detectAndHandleTimeout(long heartBeatTimeoutMs) {
+        HeartbeatTimeout heartbeatTimeout = detectTimeout(heartBeatTimeoutMs);
+
+        if (heartbeatTimeout == YES) {
+            future.complete(NO_HEARTBEAT);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    HeartbeatTimeout detectTimeout(long heartBeatTimeoutMs) {
+        if (pendingResponse != VOID) {
+            // if there is a response, then we won't timeout.
+            return NO__RESPONSE_AVAILABLE;
+        }
+
+        long currentTimeMs = currentTimeMillis();
```

Comment:
Same here, it's ambiguous if `System` or `Clock` is used as source (and which one should be used).